// @angular/core/schematics/migrations/move-document/document_import_visitor.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/migrations/move-document/document_import_visitor", ["require", "exports", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.DocumentImportVisitor = exports.DOCUMENT_TOKEN_NAME = exports.PLATFORM_BROWSER_IMPORT = exports.COMMON_IMPORT = void 0;
    const ts = require("typescript");
    exports.COMMON_IMPORT = '@angular/common';
    exports.PLATFORM_BROWSER_IMPORT = '@angular/platform-browser';
    exports.DOCUMENT_TOKEN_NAME = 'DOCUMENT';
    /** Visitor that can be used to find a set of imports in a TypeScript file. */
    class DocumentImportVisitor {
        constructor(typeChecker) {
            this.typeChecker = typeChecker;
            this.importsMap = new Map();
        }
        visitNode(node) {
            if (ts.isNamedImports(node)) {
                this.visitNamedImport(node);
            }
            ts.forEachChild(node, node => this.visitNode(node));
        }
        visitNamedImport(node) {
            if (!node.elements || !node.elements.length) {
                return;
            }
            const importDeclaration = node.parent.parent;
            // If this is not a StringLiteral it will be a grammar error
            const moduleSpecifier = importDeclaration.moduleSpecifier;
            const sourceFile = node.getSourceFile();
            let imports = this.importsMap.get(sourceFile);
            if (!imports) {
                imports = {
                    platformBrowserImport: null,
                    commonImport: null,
                    documentElement: null,
                };
            }
            if (moduleSpecifier.text === exports.PLATFORM_BROWSER_IMPORT) {
                const documentElement = this.getDocumentElement(node);
                if (documentElement) {
                    imports.platformBrowserImport = node;
                    imports.documentElement = documentElement;
                }
            }
            else if (moduleSpecifier.text === exports.COMMON_IMPORT) {
                imports.commonImport = node;
            }
            else {
                return;
            }
            this.importsMap.set(sourceFile, imports);
        }
        getDocumentElement(node) {
            const elements = node.elements;
            return elements.find(el => (el.propertyName || el.name).escapedText === exports.DOCUMENT_TOKEN_NAME);
        }
    }
    exports.DocumentImportVisitor = DocumentImportVisitor;
});
