// @angular/core/schematics/migrations/template-var-assignment/angular/html_variable_assignment_visitor.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/migrations/template-var-assignment/angular/html_variable_assignment_visitor", ["require", "exports", "@angular/compiler", "@angular/compiler/src/render3/r3_ast"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.HtmlVariableAssignmentVisitor = void 0;
    const compiler_1 = require("@angular/compiler");
    const r3_ast_1 = require("@angular/compiler/src/render3/r3_ast");
    /**
     * HTML AST visitor that traverses the Render3 HTML AST in order to find all
     * expressions that write to local template variables within bound events.
     */
    class HtmlVariableAssignmentVisitor extends r3_ast_1.NullVisitor {
        constructor() {
            super(...arguments);
            this.variableAssignments = [];
            this.currentVariables = [];
            this.expressionAstVisitor = new ExpressionAstVisitor(this.variableAssignments, this.currentVariables);
        }
        visitElement(element) {
            r3_ast_1.visitAll(this, element.outputs);
            r3_ast_1.visitAll(this, element.children);
        }
        visitTemplate(template) {
            // Keep track of the template variables which can be accessed by the template
            // child nodes through the implicit receiver.
            this.currentVariables.push(...template.variables);
            // Visit all children of the template. The template proxies the outputs of the
            // immediate child elements, so we just ignore outputs on the "Template" in order
            // to not visit similar bound events twice.
            r3_ast_1.visitAll(this, template.children);
            // Remove all previously added variables since all children that could access
            // these have been visited already.
            template.variables.forEach(v => {
                const variableIdx = this.currentVariables.indexOf(v);
                if (variableIdx !== -1) {
                    this.currentVariables.splice(variableIdx, 1);
                }
            });
        }
        visitBoundEvent(node) {
            node.handler.visit(this.expressionAstVisitor, node.handlerSpan);
        }
    }
    exports.HtmlVariableAssignmentVisitor = HtmlVariableAssignmentVisitor;
    /** AST visitor that resolves all variable assignments within a given expression AST. */
    class ExpressionAstVisitor extends compiler_1.RecursiveAstVisitor {
        constructor(variableAssignments, currentVariables) {
            super();
            this.variableAssignments = variableAssignments;
            this.currentVariables = currentVariables;
        }
        visitPropertyWrite(node, span) {
            if (node.receiver instanceof compiler_1.ImplicitReceiver &&
                this.currentVariables.some(v => v.name === node.name)) {
                this.variableAssignments.push({
                    node: node,
                    start: span.start.offset,
                    end: span.end.offset,
                });
            }
            super.visitPropertyWrite(node, span);
        }
    }
});
