// @angular/core/schematics/migrations/initial-navigation/transform.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/migrations/initial-navigation/transform", ["require", "exports", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.InitialNavigationTransform = void 0;
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    const ts = require("typescript");
    class InitialNavigationTransform {
        constructor(getUpdateRecorder) {
            this.getUpdateRecorder = getUpdateRecorder;
            this.printer = ts.createPrinter();
        }
        /** Migrate the ExtraOptions#InitialNavigation property assignments. */
        migrateInitialNavigationAssignments(literals) {
            literals.forEach(l => this.migrateAssignment(l));
        }
        /** Migrate an ExtraOptions#InitialNavigation expression to use the new options format. */
        migrateAssignment(assignment) {
            const newInitializer = getUpdatedInitialNavigationValue(assignment.initializer);
            if (newInitializer) {
                const newAssignment = ts.updatePropertyAssignment(assignment, assignment.name, newInitializer);
                this._updateNode(assignment, newAssignment);
            }
        }
        _updateNode(node, newNode) {
            const newText = this.printer.printNode(ts.EmitHint.Unspecified, newNode, node.getSourceFile());
            const recorder = this.getUpdateRecorder(node.getSourceFile());
            recorder.updateNode(node, newText);
        }
    }
    exports.InitialNavigationTransform = InitialNavigationTransform;
    /**
     * Updates the deprecated initialNavigation options to their v10 equivalents
     * (or as close as we can get).
     * @param initializer the old initializer to update
     */
    function getUpdatedInitialNavigationValue(initializer) {
        const oldText = ts.isStringLiteralLike(initializer) ?
            initializer.text :
            initializer.kind === ts.SyntaxKind.TrueKeyword;
        let newText;
        switch (oldText) {
            case false:
            case 'legacy_disabled':
                newText = 'disabled';
                break;
            case true:
            case 'legacy_enabled':
                newText = 'enabledNonBlocking';
                break;
        }
        return !!newText ? ts.createIdentifier(`'${newText}'`) : null;
    }
});
